File: src/ocaml-output/FStar_Syntax_MutRecTy.ml
open Prims
let (disentangle_abbrevs_from_bundle :
  FStar_Syntax_Syntax.sigelt Prims.list ->
    FStar_Syntax_Syntax.qualifier Prims.list ->
      FStar_Ident.lident Prims.list ->
        FStar_Range.range ->
          (FStar_Syntax_Syntax.sigelt * FStar_Syntax_Syntax.sigelt
            Prims.list))
  =
  fun sigelts ->
    fun quals ->
      fun members ->
        fun rng ->
          let sigattrs =
            FStar_List.collect (fun s -> s.FStar_Syntax_Syntax.sigattrs)
              sigelts in
          let type_abbrev_sigelts =
            FStar_All.pipe_right sigelts
              (FStar_List.collect
                 (fun x ->
                    match x.FStar_Syntax_Syntax.sigel with
                    | FStar_Syntax_Syntax.Sig_let
                        ((false,
                          {
                            FStar_Syntax_Syntax.lbname = FStar_Pervasives.Inr
                              uu___;
                            FStar_Syntax_Syntax.lbunivs = uu___1;
                            FStar_Syntax_Syntax.lbtyp = uu___2;
                            FStar_Syntax_Syntax.lbeff = uu___3;
                            FStar_Syntax_Syntax.lbdef = uu___4;
                            FStar_Syntax_Syntax.lbattrs = uu___5;
                            FStar_Syntax_Syntax.lbpos = uu___6;_}::[]),
                         uu___7)
                        -> [x]
                    | FStar_Syntax_Syntax.Sig_let (uu___, uu___1) ->
                        failwith
                          "mutrecty: disentangle_abbrevs_from_bundle: type_abbrev_sigelts: impossible"
                    | uu___ -> [])) in
          match type_abbrev_sigelts with
          | [] ->
              ({
                 FStar_Syntax_Syntax.sigel =
                   (FStar_Syntax_Syntax.Sig_bundle (sigelts, members));
                 FStar_Syntax_Syntax.sigrng = rng;
                 FStar_Syntax_Syntax.sigquals = quals;
                 FStar_Syntax_Syntax.sigmeta =
                   FStar_Syntax_Syntax.default_sigmeta;
                 FStar_Syntax_Syntax.sigattrs = sigattrs;
                 FStar_Syntax_Syntax.sigopts = FStar_Pervasives_Native.None
               }, [])
          | uu___ ->
              let type_abbrevs =
                FStar_All.pipe_right type_abbrev_sigelts
                  (FStar_List.map
                     (fun x ->
                        match x.FStar_Syntax_Syntax.sigel with
                        | FStar_Syntax_Syntax.Sig_let
                            ((uu___1,
                              {
                                FStar_Syntax_Syntax.lbname =
                                  FStar_Pervasives.Inr fv;
                                FStar_Syntax_Syntax.lbunivs = uu___2;
                                FStar_Syntax_Syntax.lbtyp = uu___3;
                                FStar_Syntax_Syntax.lbeff = uu___4;
                                FStar_Syntax_Syntax.lbdef = uu___5;
                                FStar_Syntax_Syntax.lbattrs = uu___6;
                                FStar_Syntax_Syntax.lbpos = uu___7;_}::[]),
                             uu___8)
                            ->
                            (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                        | uu___1 ->
                            failwith
                              "mutrecty: disentangle_abbrevs_from_bundle: type_abbrevs: impossible")) in
              let unfolded_type_abbrevs =
                let rev_unfolded_type_abbrevs = FStar_Util.mk_ref [] in
                let in_progress = FStar_Util.mk_ref [] in
                let not_unfolded_yet = FStar_Util.mk_ref type_abbrev_sigelts in
                let remove_not_unfolded lid =
                  let uu___1 =
                    let uu___2 = FStar_ST.op_Bang not_unfolded_yet in
                    FStar_All.pipe_right uu___2
                      (FStar_List.filter
                         (fun x ->
                            match x.FStar_Syntax_Syntax.sigel with
                            | FStar_Syntax_Syntax.Sig_let
                                ((uu___3,
                                  {
                                    FStar_Syntax_Syntax.lbname =
                                      FStar_Pervasives.Inr fv;
                                    FStar_Syntax_Syntax.lbunivs = uu___4;
                                    FStar_Syntax_Syntax.lbtyp = uu___5;
                                    FStar_Syntax_Syntax.lbeff = uu___6;
                                    FStar_Syntax_Syntax.lbdef = uu___7;
                                    FStar_Syntax_Syntax.lbattrs = uu___8;
                                    FStar_Syntax_Syntax.lbpos = uu___9;_}::[]),
                                 uu___10)
                                ->
                                let uu___11 =
                                  FStar_Ident.lid_equals lid
                                    (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v in
                                Prims.op_Negation uu___11
                            | uu___3 -> true)) in
                  FStar_ST.op_Colon_Equals not_unfolded_yet uu___1 in
                let rec unfold_abbrev_fv t fv =
                  let replacee x =
                    match x.FStar_Syntax_Syntax.sigel with
                    | FStar_Syntax_Syntax.Sig_let
                        ((uu___1,
                          {
                            FStar_Syntax_Syntax.lbname = FStar_Pervasives.Inr
                              fv';
                            FStar_Syntax_Syntax.lbunivs = uu___2;
                            FStar_Syntax_Syntax.lbtyp = uu___3;
                            FStar_Syntax_Syntax.lbeff = uu___4;
                            FStar_Syntax_Syntax.lbdef = uu___5;
                            FStar_Syntax_Syntax.lbattrs = uu___6;
                            FStar_Syntax_Syntax.lbpos = uu___7;_}::[]),
                         uu___8)
                        when
                        FStar_Ident.lid_equals
                          (fv'.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                          (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                        -> FStar_Pervasives_Native.Some x
                    | uu___1 -> FStar_Pervasives_Native.None in
                  let replacee_term x =
                    match replacee x with
                    | FStar_Pervasives_Native.Some
                        {
                          FStar_Syntax_Syntax.sigel =
                            FStar_Syntax_Syntax.Sig_let
                            ((uu___1,
                              { FStar_Syntax_Syntax.lbname = uu___2;
                                FStar_Syntax_Syntax.lbunivs = uu___3;
                                FStar_Syntax_Syntax.lbtyp = uu___4;
                                FStar_Syntax_Syntax.lbeff = uu___5;
                                FStar_Syntax_Syntax.lbdef = tm;
                                FStar_Syntax_Syntax.lbattrs = uu___6;
                                FStar_Syntax_Syntax.lbpos = uu___7;_}::[]),
                             uu___8);
                          FStar_Syntax_Syntax.sigrng = uu___9;
                          FStar_Syntax_Syntax.sigquals = uu___10;
                          FStar_Syntax_Syntax.sigmeta = uu___11;
                          FStar_Syntax_Syntax.sigattrs = uu___12;
                          FStar_Syntax_Syntax.sigopts = uu___13;_}
                        -> FStar_Pervasives_Native.Some tm
                    | uu___1 -> FStar_Pervasives_Native.None in
                  let uu___1 =
                    let uu___2 = FStar_ST.op_Bang rev_unfolded_type_abbrevs in
                    FStar_Util.find_map uu___2 replacee_term in
                  match uu___1 with
                  | FStar_Pervasives_Native.Some x -> x
                  | FStar_Pervasives_Native.None ->
                      let uu___2 =
                        FStar_Util.find_map type_abbrev_sigelts replacee in
                      (match uu___2 with
                       | FStar_Pervasives_Native.Some se ->
                           let uu___3 =
                             let uu___4 = FStar_ST.op_Bang in_progress in
                             FStar_List.existsb
                               (fun x ->
                                  FStar_Ident.lid_equals x
                                    (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v)
                               uu___4 in
                           if uu___3
                           then
                             let msg =
                               let uu___4 =
                                 FStar_Ident.string_of_lid
                                   (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v in
                               FStar_Util.format1
                                 "Cycle on %s in mutually recursive type abbreviations"
                                 uu___4 in
                             let uu___4 =
                               FStar_Ident.range_of_lid
                                 (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v in
                             FStar_Errors.raise_error
                               (FStar_Errors.Fatal_CycleInRecTypeAbbreviation,
                                 msg) uu___4
                           else unfold_abbrev se
                       | uu___3 -> t)
                and unfold_abbrev x =
                  match x.FStar_Syntax_Syntax.sigel with
                  | FStar_Syntax_Syntax.Sig_let ((false, lb::[]), uu___1) ->
                      let quals1 =
                        FStar_All.pipe_right x.FStar_Syntax_Syntax.sigquals
                          (FStar_List.filter
                             (fun uu___2 ->
                                match uu___2 with
                                | FStar_Syntax_Syntax.Noeq -> false
                                | uu___3 -> true)) in
                      let lid =
                        match lb.FStar_Syntax_Syntax.lbname with
                        | FStar_Pervasives.Inr fv ->
                            (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                        | uu___2 ->
                            failwith
                              "mutrecty: disentangle_abbrevs_from_bundle: rename_abbrev: lid: impossible" in
                      ((let uu___3 =
                          let uu___4 = FStar_ST.op_Bang in_progress in lid ::
                            uu___4 in
                        FStar_ST.op_Colon_Equals in_progress uu___3);
                       (match () with
                        | () ->
                            (remove_not_unfolded lid;
                             (match () with
                              | () ->
                                  let ty' =
                                    FStar_Syntax_InstFV.inst unfold_abbrev_fv
                                      lb.FStar_Syntax_Syntax.lbtyp in
                                  let tm' =
                                    FStar_Syntax_InstFV.inst unfold_abbrev_fv
                                      lb.FStar_Syntax_Syntax.lbdef in
                                  let lb' =
                                    let uu___4 = lb in
                                    {
                                      FStar_Syntax_Syntax.lbname =
                                        (uu___4.FStar_Syntax_Syntax.lbname);
                                      FStar_Syntax_Syntax.lbunivs =
                                        (uu___4.FStar_Syntax_Syntax.lbunivs);
                                      FStar_Syntax_Syntax.lbtyp = ty';
                                      FStar_Syntax_Syntax.lbeff =
                                        (uu___4.FStar_Syntax_Syntax.lbeff);
                                      FStar_Syntax_Syntax.lbdef = tm';
                                      FStar_Syntax_Syntax.lbattrs =
                                        (uu___4.FStar_Syntax_Syntax.lbattrs);
                                      FStar_Syntax_Syntax.lbpos =
                                        (uu___4.FStar_Syntax_Syntax.lbpos)
                                    } in
                                  let sigelt' =
                                    FStar_Syntax_Syntax.Sig_let
                                      ((false, [lb']), [lid]) in
                                  ((let uu___5 =
                                      let uu___6 =
                                        FStar_ST.op_Bang
                                          rev_unfolded_type_abbrevs in
                                      (let uu___7 = x in
                                       {
                                         FStar_Syntax_Syntax.sigel = sigelt';
                                         FStar_Syntax_Syntax.sigrng =
                                           (uu___7.FStar_Syntax_Syntax.sigrng);
                                         FStar_Syntax_Syntax.sigquals =
                                           quals1;
                                         FStar_Syntax_Syntax.sigmeta =
                                           (uu___7.FStar_Syntax_Syntax.sigmeta);
                                         FStar_Syntax_Syntax.sigattrs =
                                           (uu___7.FStar_Syntax_Syntax.sigattrs);
                                         FStar_Syntax_Syntax.sigopts =
                                           (uu___7.FStar_Syntax_Syntax.sigopts)
                                       }) :: uu___6 in
                                    FStar_ST.op_Colon_Equals
                                      rev_unfolded_type_abbrevs uu___5);
                                   (match () with
                                    | () ->
                                        ((let uu___6 =
                                            let uu___7 =
                                              FStar_ST.op_Bang in_progress in
                                            FStar_List.tl uu___7 in
                                          FStar_ST.op_Colon_Equals
                                            in_progress uu___6);
                                         (match () with | () -> tm'))))))))
                  | uu___1 ->
                      failwith
                        "mutrecty: disentangle_abbrevs_from_bundle: rename_abbrev: impossible" in
                let rec aux uu___1 =
                  let uu___2 = FStar_ST.op_Bang not_unfolded_yet in
                  match uu___2 with
                  | x::uu___3 -> let _unused = unfold_abbrev x in aux ()
                  | uu___3 ->
                      let uu___4 = FStar_ST.op_Bang rev_unfolded_type_abbrevs in
                      FStar_List.rev uu___4 in
                aux () in
              let filter_out_type_abbrevs l =
                FStar_List.filter
                  (fun lid ->
                     FStar_List.for_all
                       (fun lid' ->
                          let uu___1 = FStar_Ident.lid_equals lid lid' in
                          Prims.op_Negation uu___1) type_abbrevs) l in
              let inductives_with_abbrevs_unfolded =
                let find_in_unfolded fv =
                  FStar_Util.find_map unfolded_type_abbrevs
                    (fun x ->
                       match x.FStar_Syntax_Syntax.sigel with
                       | FStar_Syntax_Syntax.Sig_let
                           ((uu___1,
                             {
                               FStar_Syntax_Syntax.lbname =
                                 FStar_Pervasives.Inr fv';
                               FStar_Syntax_Syntax.lbunivs = uu___2;
                               FStar_Syntax_Syntax.lbtyp = uu___3;
                               FStar_Syntax_Syntax.lbeff = uu___4;
                               FStar_Syntax_Syntax.lbdef = tm;
                               FStar_Syntax_Syntax.lbattrs = uu___5;
                               FStar_Syntax_Syntax.lbpos = uu___6;_}::[]),
                            uu___7)
                           when
                           FStar_Ident.lid_equals
                             (fv'.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                             (fv.FStar_Syntax_Syntax.fv_name).FStar_Syntax_Syntax.v
                           -> FStar_Pervasives_Native.Some tm
                       | uu___1 -> FStar_Pervasives_Native.None) in
                let unfold_fv t fv =
                  let uu___1 = find_in_unfolded fv in
                  match uu___1 with
                  | FStar_Pervasives_Native.Some t' -> t'
                  | uu___2 -> t in
                let unfold_in_sig x =
                  match x.FStar_Syntax_Syntax.sigel with
                  | FStar_Syntax_Syntax.Sig_inductive_typ
                      (lid, univs, bnd, ty, mut, dc) ->
                      let bnd' =
                        FStar_Syntax_InstFV.inst_binders unfold_fv bnd in
                      let ty' = FStar_Syntax_InstFV.inst unfold_fv ty in
                      let mut' = filter_out_type_abbrevs mut in
                      [(let uu___1 = x in
                        {
                          FStar_Syntax_Syntax.sigel =
                            (FStar_Syntax_Syntax.Sig_inductive_typ
                               (lid, univs, bnd', ty', mut', dc));
                          FStar_Syntax_Syntax.sigrng =
                            (uu___1.FStar_Syntax_Syntax.sigrng);
                          FStar_Syntax_Syntax.sigquals =
                            (uu___1.FStar_Syntax_Syntax.sigquals);
                          FStar_Syntax_Syntax.sigmeta =
                            (uu___1.FStar_Syntax_Syntax.sigmeta);
                          FStar_Syntax_Syntax.sigattrs =
                            (uu___1.FStar_Syntax_Syntax.sigattrs);
                          FStar_Syntax_Syntax.sigopts =
                            (uu___1.FStar_Syntax_Syntax.sigopts)
                        })]
                  | FStar_Syntax_Syntax.Sig_datacon
                      (lid, univs, ty, res, npars, mut) ->
                      let ty' = FStar_Syntax_InstFV.inst unfold_fv ty in
                      let mut' = filter_out_type_abbrevs mut in
                      [(let uu___1 = x in
                        {
                          FStar_Syntax_Syntax.sigel =
                            (FStar_Syntax_Syntax.Sig_datacon
                               (lid, univs, ty', res, npars, mut'));
                          FStar_Syntax_Syntax.sigrng =
                            (uu___1.FStar_Syntax_Syntax.sigrng);
                          FStar_Syntax_Syntax.sigquals =
                            (uu___1.FStar_Syntax_Syntax.sigquals);
                          FStar_Syntax_Syntax.sigmeta =
                            (uu___1.FStar_Syntax_Syntax.sigmeta);
                          FStar_Syntax_Syntax.sigattrs =
                            (uu___1.FStar_Syntax_Syntax.sigattrs);
                          FStar_Syntax_Syntax.sigopts =
                            (uu___1.FStar_Syntax_Syntax.sigopts)
                        })]
                  | FStar_Syntax_Syntax.Sig_let (uu___1, uu___2) -> []
                  | uu___1 ->
                      failwith
                        "mutrecty: inductives_with_abbrevs_unfolded: unfold_in_sig: impossible" in
                FStar_List.collect unfold_in_sig sigelts in
              let new_members = filter_out_type_abbrevs members in
              let new_bundle =
                {
                  FStar_Syntax_Syntax.sigel =
                    (FStar_Syntax_Syntax.Sig_bundle
                       (inductives_with_abbrevs_unfolded, new_members));
                  FStar_Syntax_Syntax.sigrng = rng;
                  FStar_Syntax_Syntax.sigquals = quals;
                  FStar_Syntax_Syntax.sigmeta =
                    FStar_Syntax_Syntax.default_sigmeta;
                  FStar_Syntax_Syntax.sigattrs = sigattrs;
                  FStar_Syntax_Syntax.sigopts = FStar_Pervasives_Native.None
                } in
              (new_bundle, unfolded_type_abbrevs)

File: src/syntax/FStar.Syntax.MutRecTy.fs
(*
   Copyright 2008-2014 Nikhil Swamy and Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
#light "off"
module FStar.Syntax.MutRecTy
open FStar.Syntax.Syntax
open FStar.Ident
open FStar.Util
open FStar.Errors
open FStar.Syntax.InstFV
module S = FStar.Syntax.Syntax
module SS = FStar.Syntax.Subst
module U = FStar.Util


// VALS_HACK_HERE

(* Given a list of bundled type declarations potentially with type
   abbreviations, construct the new bundle without type abbreviations
   or lets (where they have been all unfolded) and the list of type
   abbreviations or lets separated away from the bundle (and sorted in
   dependency order, in such a way that they are no longer mutually
   recursive.)  *)

let disentangle_abbrevs_from_bundle
    (sigelts: list<sigelt>)
    (quals:   list<qualifier>)
    (members: list<lident>)
    (rng:   FStar.Range.range)
    : sigelt * list<sigelt> =

    (* Gather the list of type abbrevs *)
   let type_abbrev_sigelts = sigelts |> List.collect begin fun x -> match x with
       | Sig_let ((false, [ { lbname= Inr _ } ]), _, _, _, _) -> [x]
       | Sig_let (_, _, _, _, _) ->
         failwith "mutrecty: disentangle_abbrevs_from_bundle: type_abbrev_sigelts: impossible"
       | _ -> []
   end
   in

   match type_abbrev_sigelts with
   | [] ->
     (* if there are no type abbreviations, then do not change anything. *)
     (Sig_bundle (sigelts, quals, members, rng), [])
   | _ ->

    let type_abbrevs = type_abbrev_sigelts |> List.map begin function
        | Sig_let ((_, [ { lbname = Inr fv } ] ) , _, _, _, _) -> fv.fv_name.v
        | _ -> failwith "mutrecty: disentangle_abbrevs_from_bundle: type_abbrevs: impossible"
    end
    in

    (* First, unfold type abbrevs among themselves *)
    let unfolded_type_abbrevs =

        (* List of type abbreviations that have been unfolded, in
        reverse order (from most recent to most ancient: the head
        depends on the tail.) *)
        let rev_unfolded_type_abbrevs : ref<(list<sigelt>)> = U.mk_ref [] in

        (* List of names of type abbreviations whose unfolding has
        started. If they occur during renaming of the current type
        abbreviation, then there is a cycle. Follows a stack
        discipline. *)
        let in_progress : ref<(list<lident>)> = U.mk_ref [] in

        (* List of type abbreviations that have not been unfolded
        yet. Their order can change, since anyway they will be
        reordered after being unfolded. *)
        let not_unfolded_yet = U.mk_ref type_abbrev_sigelts in

        let remove_not_unfolded lid =
            not_unfolded_yet := !not_unfolded_yet |> List.filter begin function
                | Sig_let ((_, [ { lbname = Inr fv } ] ) , _, _, _, _) ->
                  not (lid_equals lid fv.fv_name.v)
                | _ -> true
            end
        in
    
        (* Replace a free variable corresponding to a type
        abbreviation, with memoization. *)
        let rec unfold_abbrev_fv (t: term) (fv : S.fv) : term =
            let replacee x = match x with
                | Sig_let ((_, [ { lbname = Inr fv' } ] ) , _, _, _, _)
                  when lid_equals fv'.fv_name.v fv.fv_name.v ->
                  Some x
                | _ -> None
            in
            let replacee_term x = match replacee x with
                | Some (Sig_let ((_, [ { lbdef = tm } ] ) , _, _, _, _)) -> Some tm
                | _ -> None
            in
            match U.find_map !rev_unfolded_type_abbrevs replacee_term with
                | Some x -> x
                | None ->
                  begin match U.find_map type_abbrev_sigelts replacee with
                      | Some se ->
                          if FStar.List.existsb (fun x -> lid_equals x fv.fv_name.v) !in_progress
                          then let msg = U.format1 "Cycle on %s in mutually recursive type abbreviations" fv.fv_name.v.str in
                               raise (Error (msg, range_of_lid fv.fv_name.v))
                          else unfold_abbrev se
                      | _ -> t
                  end

        (* Start unfolding in a type abbreviation that has not occurred before. *)
        and unfold_abbrev = function
            | Sig_let ((false, [lb]) , rng, _, quals, attr) ->
	        (* eliminate some qualifiers for definitions *)
	        let quals = quals |> List.filter begin function
	        | Noeq -> false
	        | _ -> true
	        end in
                let lid = match lb.lbname with
                    | Inr fv -> fv.fv_name.v
                    | _ -> failwith "mutrecty: disentangle_abbrevs_from_bundle: rename_abbrev: lid: impossible"
                in
                let () = in_progress := lid :: !in_progress in  (* push *)
                let () = remove_not_unfolded lid in
                let ty' = inst unfold_abbrev_fv lb.lbtyp in
                let tm' = inst unfold_abbrev_fv lb.lbdef in
                let lb' = { lb with lbtyp = ty' ; lbdef = tm' } in
                let sigelt' = Sig_let ((false, [lb']), rng, [lid], quals, attr) in
                let () = rev_unfolded_type_abbrevs := sigelt' :: !rev_unfolded_type_abbrevs in
                let () = in_progress := List.tl !in_progress in (* pop *)
                tm'
            | _ -> failwith "mutrecty: disentangle_abbrevs_from_bundle: rename_abbrev: impossible"
        in
            
        let rec aux () = match !not_unfolded_yet with
            | x :: _ -> let _unused = unfold_abbrev x in aux ()
            | _ -> List.rev !rev_unfolded_type_abbrevs

        in

        aux ()
      in

      (* Now, unfold in inductive types and data constructors *)

      let filter_out_type_abbrevs l =
          List.filter (fun lid -> FStar.List.for_all (fun lid' -> not (lid_equals lid lid')) type_abbrevs) l
      in
      
      let inductives_with_abbrevs_unfolded =

          let find_in_unfolded fv = U.find_map unfolded_type_abbrevs begin fun x -> match x with
              | Sig_let ((_, [ { lbname = Inr fv' ; lbdef = tm } ] ), _, _, _, _) when (lid_equals fv'.fv_name.v fv.fv_name.v) ->
                Some tm
              | _ -> None
          end
          in

          let unfold_fv (t: term) (fv: S.fv) : term = match find_in_unfolded fv with
              | Some t' -> t'
              | _ -> t
          in

          let unfold_in_sig = function
              | Sig_inductive_typ (lid, univs, bnd, ty, mut, dc, quals, rng) ->
                let bnd' = inst_binders unfold_fv bnd in
                let ty'  = inst unfold_fv ty in
                let mut' = filter_out_type_abbrevs mut in
                [Sig_inductive_typ (lid, univs, bnd', ty', mut', dc, quals, rng)]

              | Sig_datacon (lid, univs, ty, res, npars, quals, mut, rng) ->
                let ty' = inst unfold_fv ty in
                let mut' = filter_out_type_abbrevs mut in
                [Sig_datacon (lid, univs, ty', res, npars, quals, mut', rng)]

              | Sig_let (_, _, _, _, _) ->
                []

              | _ -> failwith "mutrecty: inductives_with_abbrevs_unfolded: unfold_in_sig: impossible"
           in

           List.collect unfold_in_sig sigelts
      in

      (* Finally, construct a new bundle separate from type abbrevs *)

      let new_members = filter_out_type_abbrevs members
      in

      let new_bundle = Sig_bundle (inductives_with_abbrevs_unfolded, quals, new_members, rng)
      in

      (new_bundle, unfolded_type_abbrevs)
